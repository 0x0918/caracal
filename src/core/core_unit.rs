use crate::detectors::detector::Result;
use crate::detectors::get_detectors;
use crate::{Args, CompilationUnit};
use crate::{Filter, Print};

pub struct CoreUnit<'a> {
    compilation_unit: CompilationUnit<'a>,
    args: Args,
}

impl<'a> CoreUnit<'a> {
    pub fn new(compilation_unit: CompilationUnit<'a>, args: Args) -> Self {
        CoreUnit {
            compilation_unit,
            args,
        }
    }

    pub fn get_compilation_unit(&self) -> &CompilationUnit {
        &self.compilation_unit
    }

    pub fn run(&mut self) {
        self.compilation_unit.analyze();
        let mut results: Vec<Result> = Vec::new();
        let detectors_to_run = get_detectors();
        for d in detectors_to_run {
            results.extend(d.run(self));
        }

        for r in results {
            println!(
                "{} impact: {} confidence: {}\n {}",
                r.name, r.impact, r.confidence, r.message
            );
        }

        self.run_printer();
    }

    fn run_printer(&self) {
        match self.args.filter {
            Filter::All => match self.args.print {
                Some(Print::Cfg) => {
                    for f in self.compilation_unit.functions() {
                        f.cfg_to_dot(f.get_cfg());
                    }
                }
                Some(Print::CfgOptimized) => {
                    for f in self.compilation_unit.functions() {
                        f.cfg_to_dot(f.get_cfg_optimized());
                    }
                }
                _ => (),
            },
            Filter::UserFunctions => match self.args.print {
                Some(Print::Cfg) => {
                    for f in self.compilation_unit.functions_user_defined() {
                        f.cfg_to_dot(f.get_cfg());
                    }
                }
                Some(Print::CfgOptimized) => {
                    for f in self.compilation_unit.functions_user_defined() {
                        f.cfg_to_dot(f.get_cfg_optimized());
                    }
                }
                _ => (),
            },
        }
    }
}
