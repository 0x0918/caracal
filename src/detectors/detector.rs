use crate::core::core_unit::CoreUnit;

use std::fmt;

pub trait Detector {
    fn name(&self) -> &str;
    fn impact(&self) -> &Impact;
    fn confidence(&self) -> &Confidence;
    fn run(&self, core: &CoreUnit) -> Vec<Result>;
}

#[derive(Debug, Clone, Copy)]
pub enum Impact {
    High,
    Medium,
    Low,
    Informational,
}

impl fmt::Display for Impact {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Impact::High => write!(f, "High"),
            Impact::Medium => write!(f, "Medium"),
            Impact::Low => write!(f, "Low"),
            Impact::Informational => write!(f, "Informational"),
        }
    }
}

#[derive(Debug, Clone, Copy)]
pub enum Confidence {
    High,
    Medium,
    Low,
}

impl fmt::Display for Confidence {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Confidence::High => write!(f, "High"),
            Confidence::Medium => write!(f, "Medium"),
            Confidence::Low => write!(f, "Low"),
        }
    }
}

pub struct Result {
    pub name: String,
    pub impact: Impact,
    pub confidence: Confidence,
    pub message: String,
}
