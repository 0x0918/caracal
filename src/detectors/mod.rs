use self::detector::Detector;

pub mod controlled_library_call;
pub mod detector;
pub mod unused_events;

pub fn get_detectors() -> Vec<Box<dyn Detector>> {
    vec![
        Box::<controlled_library_call::ControlledLibraryCall>::default(),
        Box::<unused_events::UnusedEvents>::default(),
    ]
}
