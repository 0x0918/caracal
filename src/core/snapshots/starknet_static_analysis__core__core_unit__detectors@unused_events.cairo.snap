---
source: src/core/core_unit.rs
expression: "get_detectors().iter().flat_map(|d| d.run(&core)).collect::<Vec<Result>>()"
---
[
    Result {
        name: "unused-events",
        impact: Medium,
        confidence: Medium,
        message: "Event MyUnusedEvent defined in unused_events::unused_events::UnusedEvents is never emitted",
    },
]
